plugins/externalwrench/include/gazebo/ExternalWrench.hh

```diff
@@ -0,0 +1,83 @@
+#ifndef EXTERNALWRENCH_H
+#define EXTERNALWRENCH_H
+
+#include <iostream>
+#include <stdlib.h>
+#include <ctime>
+#include <stdio.h>
+#include <string>
+#include <memory>
+
+#include "gazebo/gazebo.hh"
+#include <gazebo/physics/PhysicsEngine.hh>
+#include <gazebo/common/Events.hh>
+#include <gazebo/common/Plugin.hh>
+#include <gazebo/transport/Node.hh>
+#include <gazebo/physics/Link.hh>
+#include <gazebo/physics/World.hh>
+#include <gazebo/physics/Model.hh>
+#include <GazeboYarpPlugins/common.h>
+
+#include <yarp/os/Network.h>
+#include <yarp/os/RpcServer.h>
+#include <yarp/os/Bottle.h>
+#include <yarp/sig/Vector.h>
+#include <yarp/os/Thread.h>
+#include <yarp/os/Time.h>
+#include <yarp/os/Vocab.h>
+#include <yarp/os/Log.h>
+#include <yarp/os/LogStream.h>
+#include <yarp/math/Math.h>
+
+#include <boost/lexical_cast.hpp>
+
+using namespace gazebo;
+
+class ExternalWrench
+{
+private:
+
+   static int count;
+   float      color[4];
+   struct wrenchCommand
+   {
+        std::string              link_name;
+        yarp::sig::Vector        force;
+        yarp::sig::Vector        torque;
+        double                   duration;
+   };
+
+   std::unique_ptr<wrenchCommand> wrenchPtr{new wrenchCommand()};
```

Comment:
@traversaro I tried to use without pointers 

> > On resetting gazebo, the wrenches are still applied as provided by the user commands. Do you think the wrenches should be cleared? @traversaro
> 
> My understanding of reset is that the simulation should restart from the initial condition. If a user asked for an external force of 1 N on one link for one hour, I think after a reset, the force should be cleared. Think for example of an RL scenario: would it make sense to keep the old force?

Agreed!

Reply:
> I am a bit confused. Why are you using `std::unique_ptr` here?
> Why not a simple attribute of the class:
> 
> ```c++
>   wrenchCommand wrench;
> ```
> 
> or even better, given that the structure `wrenchCommand` is not used elsewhere, an anonymous struct:
> 
> ```c++
>    struct
>    {
>         std::string              link_name;
>         yarp::sig::Vector        force;
>         yarp::sig::Vector        torque;
>         double                   duration;
>    } wrench;
> ```
> 
> or given that the structure is never used standalone, why not simply including the attributes of the structure as attributes of the class?

@traversaro I remember back in time starting out with simple data structures without pointers but then I quickly ran into some problems while handling dynamic wrench addition and removal in multiple wrench operation mode. That is why I belive I migrated to using pointer to handle them correctly. Now, I tested to change back to simple data structures again and in the multiple wrench case the visuals are funny looking 

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/6505998/55796072-628f5e00-5ac9-11e9-8886-87304d7d5dc6.gif)

But with the pointers usage it is stable

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/6505998/55796126-85217700-5ac9-11e9-989a-d331984dbe03.gif)

At the moment I prefer not to investigate it further and I would like to stick to the pointers usage.

Reply:
You obtain this behavior just by changing `std::unique_ptr<wrenchCommand> wrenchPtr{new wrenchCommand()};` to `wrenchCommand wrench;`? If that is the case, it is extremely strange and I do not know what's going  on. 

If instead you obtain  this because you changed `boost::shared_ptr<ExternalWrench> newWrench(new ExternalWrench);` to `ExternalWrench newWrench` and other changes related to `ExternalWrench`,  I guess what is going on is that there is something going on related to the use of `static count`, that is increased whenever a new copy of `ExternalWrench` is created, but not when a new copy of the shared_ptr is created. This is exactly the tricky behavior related to static variables (that are actually just global variables a bit disguised) that I  was mentioning in  https://github.com/robotology/gazebo-yarp-plugins/pull/418/files#r273442326 . 
I think the plugin will be much more maintainable if you just drop the `static count` and just  have a normal local `index` or `count` attribute that can be set with a `SetIndex` or `SetCount` method, that can contain all the code currently contained in the constructor. This `SetIndex`/`SetCount` method can be set directly in the https://github.com/robotology/gazebo-yarp-plugins/pull/418/files#diff-6f34d47dcd5558cbef158050b5563c76R175 , using some kind of  variable shared between `RPCServerThread` and  `ApplyExternalWrench` that is increase/decreased appropriately.

Reply:
This behavior is after changing `std::unique_ptr<wrenchCommand> wrenchPtr{new wrenchCommand()}; to wrenchCommand wrench;` and also `boost::shared_ptr< std::vector< boost::shared_ptr<ExternalWrench>>> wrenchesVectorPtr{new std::vector< boost::shared_ptr<ExternalWrench>>()};` to `std::vector<ExternalWrench> wrenchesVector`. Anyway, I do agree that `static count` variable will eventually cause problems when having multiple model having the plugin. So, I will have a variable `wrenchesIndex` in `ApplyExternalWrench` class and set the `wrenchIndex` variable in `ExternalWrench` class.

Reply:
`static count` is removed and updated with new wrench count and index variables https://github.com/robotology/gazebo-yarp-plugins/pull/418/commits/ada82910ef25b7c3875faffd149d8802566f211a

Shared points are removed and simple data structures are used https://github.com/robotology/gazebo-yarp-plugins/pull/418/commits/b7d1d383c7b4ed478092b07945afe13e13960b4e